### PR TITLE
added pdfium project

### PR DESCRIPTION
**Pull request by mbiggio**

### CODE REVIEW

File: gvsbuild/projects.py [hunk 1]
```diff
@@ -409,6 +409,30 @@ def build(self):
...
+    def build(self):
+        self.exec_vs(r'cp ..\resources.rc .')
+        self.exec_vs(r'cp ..\args.gn out')
+        if self.builder.x86:
+            self.exec_vs(r'echo target_cpu="x86" >> out\args.gn')
+        self.exec_vs(r'gn gen out')
+        self.exec_vs(r'ninja -C out pdfium')
+
+        self.install_dir(r'public', r'include\pdfium')
+        self.install(r'out\pdfium.dll.lib', r'lib')
```
Comment (nacho) [inline gvsbuild/projects.py:434]: I think we want this to be named pdfium.lib for consistency with other libraries

File: gvsbuild/projects.py [hunk 2]
```diff
@@ -409,6 +409,30 @@ def build(self):
...
+            dependencies = ['depot', 'python', 'ninja'],
+            patches = ['shared_library.patch', 'relative_includes.patch'],
+            subproject_patches = {'build': ['rc_compiler.patch']},
+            )
+
+    def build(self):
+        self.exec_vs(r'cp ..\resources.rc .')
+        self.exec_vs(r'cp ..\args.gn out')
+        if self.builder.x86:
+            self.exec_vs(r'echo target_cpu="x86" >> out\args.gn')
```
Comment (nacho) [inline gvsbuild/projects.py:429]: I'd rather a more pythonish way of doing this but ok...

**Review by nacho — COMMENTED** [review]
It looks good in general but let's split out the patches related to the subproject, adding the tools, the expander  and finally pdfium

**Comment by nacho** [comment]
@guruDanny67 patches look good to me. Can you double check them?

**Comment by guruDanny67** [comment]
I take a fast look and there are some notes.

Using the 'cp' command to copy the file seems not so correct to me: on my setup an error is generated (I have to add msys2 to the dependencies to continue the build) and if something goes wrong (access error, file missing, ...) I don't know if the error comes to the script or not. We should use shutil.copy(...) or shutil.copy2(...).

The expander update(...) method simply delete all the data and does an unpack() so every time a complete, full download is done: normally we try to update what we have (with git, for examples) to be fast and to avoid not necessary download: it's like that every time we build we download all the tarballs of all the projetcs, if it's avoidable is better.

I can't build the project because some sub-projects doesn't found a visual studio version suitable: I have the tools not in the standard directory (I'm an old 54 year, today btw :), programmer and you can't convince me that 'c:\Program Files (x86)\Microsoft Visual Studio\2019\coomunity' is a better dir then 'c:\dev\vs2019') but this is the lesser of the problem: I can have more then one Vs installed and the script MUST use the one I select, not something some tools found around so using the passed vs version & path we use for all the build tools lets me build all and be always correct (at some time I have vs 2015, vs 2017, vs 2019 and one build tool and mixing, without control, this can only lead to problems).

One last thing: the patches apply with warnings (offset & stripping CR): while this can happen after a while (we made the patch for version 1.0, now we build 1.5) I think that for the first commit there should be none of them.

With best regards

    Daniele Forghieri

**Comment by mbiggio** [comment]
@guruDanny67 thanks a lot for your comments on this patch.
With the new commits some of the issues you pointed out should have been fixed:
* pdfium now uses the VS version passed in input to the `build` script, instead of trying to autodetect it in the system
* the warning in the patches shouldn't be there anymore
* the `cp` command has been replaced

Also (unrelated to pdfium) NASM has been updated to a more recent version, to fix a problem we were facing while building `turbojpeg`:
```
[252/342] Building ASM_NASM object simd\CMakeFiles\simd.dir\i386\jchuff-sse2.asm.obj
FAILED: simd/CMakeFiles/simd.dir/i386/jchuff-sse2.asm.obj 
C:\gvsbuild\tools\nasm-2.13.03\nasm.exe -I. -I..\ -DWIN32 -I"C:/gvsbuild/build/Win32/debug/libjpeg-turbo/simd/nasm/" -I"C:/gvsbuild/build/Win32/debug/libjpeg-turbo/simd/i386/" -g -f win32 -o simd\CMakeFiles\simd.dir\i386\jchuff-sse2.asm.obj ..\simd\i386\jchuff-sse2.asm
```

However, this issue:
> The expander update(...) method simply delete all the data and does an unpack() so every time a complete, full download is done: normally we try to update what we have (with git, for examples) to be fast and to avoid not necessary download: it's like that every time we build we download all the tarballs of all the projetcs, if it's avoidable is better.

was not addressed, because I'm not familiar enough with how google build systems manages project syncing, so I would propose to keep it like that for the moment

**Comment by mbiggio** [comment]
In the latest push I actually removed the NASM update and opened a different PR for it

**Comment by nacho** [comment]
@guruDanny67 I think we should be ok now, but I'll let you give the OK :)

**Comment by guruDanny67** [comment]
I'll try to look at the PR this afternoon.

Thanks

Il giorno lun 21 giu 2021 alle ore 10:49 Ignacio Casal Quinteiro <
***@***.***> ha scritto:

**Comment by guruDanny67** [comment]
I don't like to be the one that founds problems but it's obviously my superpower and my curse :(

Building the project at some time the script stop with the error:

subprocess.CalledProcessError: Command 'gn gen out' returned non-zero exit status 1.

but the original error seems a little bit before:

'''
Exception: dbghelp.dll not found in "C:\Program Files (x86)\Windows Kits\10\Debuggers\x64\dbghelp.dll"
You must install Windows 10 SDK version 10.0.19041.0 including the "Debugging Tools for Windows" feature.
ERROR at //build/toolchain/win/BUILD.gn:49:3: Script returned non-zero exit code.
'''

I try to see if the Debug tools are present in the VS installer but I found nothing.

I look for the windows sdk, I download the winsdksetup, launch it and a lot of options appear. You can't get only the Debug tools so I select some more files, everything is downloaded but nothing change, same error because, I thought, the winsdksetup only downloads the file, nothing is installed and you have to install manually what you want but I don't have luck and I'm stuck at the beginning (after the download the program says that you found winsdksetup.exe in the download dir to install the sdk but also the new one only downloads the files).

So how can I proceed? 

Note: the exact instruction for having a correct build system must be explained in some docs for the project and, BTW, what is PDFium?

With best regards
    Daniele Forghieri

**Comment by mbiggio** [comment]
@guruDanny67 thanks for the time you spent testing this patch.

Maybe the reason why you cannot install the debug tools is because you already have an SDK installed on the system, maybe without the debug tools?

Could you check that from  `Control Panel` -> `Apps and Features` if you already have the SDK or not, and in case you have it, uninstall it and reinstall it, including the debug tools?

About pdfium, this is a google library used to manipulate and render PDF files:
https://pdfium.googlesource.com/pdfium/+/refs/heads/main/README.md
https://opensource.google/projects/pdfium

In the prerequisites, pdfium indeed mentions that SDK Debugging Tools are needed:
https://chromium.googlesource.com/chromium/src/+/main/docs/windows_build_instructions.md#System-requirements

**Comment by danyeaw** [comment]
@mbiggio Should we add some additional docs on how to build this to the README and then get it merged in?

**Comment by danyeaw** [comment]
This PR is now stale and I haven't received a response from you. If you are still interested in adding pdfium, please reopen it, rebase it, and add some docs on what must be installed by users.

**Pull request closed without merging**